pkg/metrics/metrics_test.go

```diff
@@ -353,3 +353,141 @@ func TestWriteMetric(t *testing.T) {
 		})
 	}
 }
+
+func TestWriteMetricAsync(t *testing.T) {
+	t.Parallel()
+
+	// Record calls made to test server. Separate per test using a per-test
+	// unique id in URL.
+	var reqMap sync.Map
+
+	// Request body is intentionally leaked to allow for inspection in test cases.
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Add 10ms of latency to response to allow testing timeout.
+		time.Sleep(10 * time.Millisecond)
+		saveReq := r.Clone(context.Background())
+		body, err := io.ReadAll(r.Body)
+		if err != nil {
+			t.Fatalf("error copying body: %s", err.Error())
+		}
+		saveReq.Body = io.NopCloser(bytes.NewBuffer(body))
+		r.Body = io.NopCloser(bytes.NewBuffer(body))
+		_, prevExist := reqMap.Swap(r.RequestURI, saveReq)
+		if prevExist {
+			t.Fatalf("multiple requests to same url: %s", r.RequestURI)
+		}
+		if !strings.HasSuffix(r.RequestURI, "/sendMetrics") {
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprintln(w, http.StatusText(http.StatusNotFound))
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, "ok")
+	}))
+
+	t.Cleanup(func() {
```

Comment:
You can also just `defer ts.Close()`. This is fine though.

Reply:
Or `t.Cleanup(ts.Close)`. `defer` won't work if you use subtests, because it'll close the server before the tests run.

Reply:
Stuck in java land, always forget to think of functions as values